Fixes #30779 - Implement ActionDispatch::RemoteIp

ekohl opened the pull request:
The ActionDispatch::RemoteIp middleware can provide the remote IP correctly as a property of the request object. It has an option to provide a list of trusted reverse proxies which needs to be exposed to the user. The installer should also expose this. It only needs to be in settings.yaml since middleware can't dynamically reload config.

The middleware is only added if the setting is present. That's because using the middleware without a reverse proxy actually opens up a security risk, as noted in its documentation.

Currently untested but posting here so I can reference it. It may be relevant to https://github.com/theforeman/foreman/pull/7909.

ekohl committed:
Fixes #30779 - Implement ActionDispatch::RemoteIp

The ActionDispatch::RemoteIp middleware can provide the remote IP
correctly as a property of the request object. This looks at
X-Forwarded-For headers while stripping out reverse proxies and falls
back to REMOTE_ADDR.

It has an option to provide a list of trusted reverse proxies which
needs to be exposed to the user. This option is exposed via a setting
which the installer can set. It needs to be in settings.yaml since
middleware can't dynamically reload config.

This middleware is enabled by default in Rails, which is an odd choice
because it actually opens up a security risk, as noted in its
documentation. Technically the best choice would be to disable the
middleware if running without a reverse proxy (like with Passenger).

It should also be noted that while there appears to be spoofing
protection inside the middleware, it doesn't actually work since the
Client-Ip is never set in our deployments.

This commit also consistently uses request.remote_ip to obtain the IP.

After this change, the remote_addr setting is no longer needed and also
no longer makes sense.

Co-authored-by: Marek Hulán <mhulan@redhat.com>

ekohl commented:
It should be noted that this may break Katello's reverse proxy. Katello deploys a reverse proxy on every content proxy but doesn't have any security around it and doesn't register them. It's likely that this middleware will raise an error.

There is no migration from the regex in the DB to a settings file. That should be an upgrade warning.

lzap commented:
Also update `settings.yaml.example` with some examples. Are these IPs? Hostnames? Does it support IPv6?

It would be nice to log a warning if reverse proxied request was ignored, that's a bonus. Overall this makes sense.

ekohl commented on config/application.rb:
```diff
@@ -200,10 +200,15 @@ class Application < Rails::Application
+    #
+    # currently the requests with untrusted IPs in X_FORWARDED_FOR are taken into the consideration,
+    # the spoof detection based on HTTP_CLIENT_IP header performed by the middleware does not work
+    config.action_dispatch.trusted_proxies = %w(127.0.0.1/8 ::1).map { |proxy| IPAddr.new(proxy) }
```
I still think a `SETTINGS[:trusted_proxies]` is good to have. It allows overriding via YAML if needed. It can default to localhost only (as you have it now).

ares replied:
then we need the installer patch which I would prefer to avoid for the stable branch

ekohl replied:
Technically no, as long as there is a hardcoded fallback. So I think that would be `SETTINGS.fetch(:trusted_proxies, %w(127.0.0.1/8 ::1))` (and then map those to IPAddr).

ares replied:
alright, I'll take a look at that tomorrow, not a big deal

ares commented:
> It should be noted that this may break Katello's reverse proxy. Katello deploys a reverse proxy on every content proxy but doesn't have any security around it and doesn't register them. It's likely that this middleware will raise an error.

As it was figured out, this is not an issue, because the spoof check never raises the exception.

We also need a migration to drop the old setting, I don't have an easy way to add file to the PR, could you please add the following as

db/migrate/20210317170111_remove_the_remote_add_setting.rb
```ruby
class RemoveTheRemoteAddSetting < ActiveRecord::Migration[6.0]
  def up
    Setting.find_by_name('remote_addr').try(:destroy)
  end

  def down
    # no action, seeding on app start will create the object with the correct default value
  end
end
```

I've tested this on production environment and all seem to be working.

ekohl commented:
I've updated this:
* Rebased
* Re-introduced the `SETTINGS[:trusted_proxies]`
* Added the DB migration
* Consistently use `request.remote_ip`
* Squashed and updated the commit message

I haven't had a chance to test it yet, but I do think it's ready for review.

ares commented:
The code looks well, I'll retest later today, if I don't find anything broken I plan to merge. Thanks @ekohl for quick update!

ares reviewed (approved):
test failed on ruby 2.6 with

```
-- C level backtrace information -------------------------------------------
/tmp/jenkins4304847768048863449.sh: line 53:  5571 Segmentation fault      bundle exec rake $tasks TESTOPTS="-v" --trace
```
2.5 and 2.7 is green, 2.6 issue is I think is unrelated, I've tested what I could including setting the custom value in settings.yaml, which work fine.

Thanks @ekohl also for updating the commit message, I'm merging this now! I'll also ask for including it in the 2.4-stable if possible.

tbrisker commented:
2.3 - a9cf7130b
